File: 65. Valid Number/Solution.py
class Solution:
    def isNumber(self, s: str) -> bool:
        n,i,num,dot,ex = len(s), (1 if s[0] in "+-" else 0), False,False,False
        while i<n:
            if s[i].isdigit():
                num = True
            elif s[i]=='.' and not dot and not ex:
                dot = True
            elif s[i] in "eE" and not ex and num:
                ex,num = True,False
                i += (1 if i+1<n and s[i+1] in "+-" else 0)
            else:
                return False
            i += 1
        return num

File: 65. Valid Number/Solution.cpp
class Solution {
public:
    bool isNumber(string s) {
        int n = s.size();
        int i = 0;
        if(s[0]=='+' || s[0]=='-') i++;
        bool num = false, dot = false, ex = false;
        for(; i<n; i++){
            if(isdigit(s[i])){
                num = true;
            }else if(s[i]=='.'){
                if(dot || ex) return false;
                dot = true;
            }else if(s[i]=='e' || s[i]=='E'){
                if(ex || !num) return false;
                ex = true;
                num = false;
                if(i+1<n && (s[i+1]=='+' || s[i+1]=='-')) i++;
            }else{
                return false;
            }
        }
        return num;
    }
};
